src/test/java/com/amazon/dlic/auth/http/jwt/keybyoidc/SelfRefreshingKeySetTest.java

```diff
@@ -11,108 +11,100 @@
+        Assert.assertEquals(TestJwk.OCT_1_K, jsonWebKey.getKeyProperty("k"));
+    }
 
-        provider.waitForCalled();
+    @Test
+    public void getKey_withNullKidShouldThrowAuthenticatorUnavailableException() throws AuthenticatorUnavailableException,
+        BadCredentialsException {
 
-        Future<JsonWebKey> f2 = executorService.submit(() -> selfRefreshingKeySet.getKey("kid/b"));
+        Assert.assertThrows(AuthenticatorUnavailableException.class, () -> selfRefreshingKeySet.getKey(null));
 
-        while (selfRefreshingKeySet.getQueuedGetCount() == 0) {
-            Thread.sleep(10);
-        }
+    }
 
-        provider.unblock();
+    @Test
+    public void getKey_withInvalidDataShouldReturnBadCredentialException() throws AuthenticatorUnavailableException,
+        BadCredentialsException {
 
-        Assert.assertEquals(TestJwk.OCT_1_K, f1.get().getProperty("k"));
-        Assert.assertEquals(TestJwk.OCT_2_K, f2.get().getProperty("k"));
+        Assert.assertThrows(BadCredentialsException.class, () -> selfRefreshingKeySet.getKey("kid/X"));
+    }
 
-        Assert.assertEquals(1, selfRefreshingKeySet.getRefreshCount());
-        Assert.assertEquals(1, selfRefreshingKeySet.getQueuedGetCount());
+    @Test
+    public void getKeyAfterRefresh_withKidShouldReturnKey() throws AuthenticatorUnavailableException, BadCredentialsException {
 
+        JsonWebKey jsonWebKey = selfRefreshingKeySet.getKeyAfterRefresh("kid/b");
+        Assert.assertEquals(TestJwk.OCT_2_K, jsonWebKey.getKeyProperty("k"));
     }
 
-    static class MockKeySetProvider implements KeySetProvider {
-
-        @Override
-        public JsonWebKeys get() throws AuthenticatorUnavailableException {
-            return TestJwk.OCT_1_2_3;
+    @Test
+    public void getKeyAfterRefresh_queuedGetCountVariableShouldBeZeroWhenFinishWithAllKeyRefreshes() throws InterruptedException,
+        ExecutionException {
+
+        int numThreads = 10;
+        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
+        Object lock = new Object();
+
+        for (int i = 0; i < numThreads; i++) {
+            executor.submit(() -> {
+                synchronized (lock) {
+                    try {
+                        selfRefreshingKeySet.getKeyAfterRefresh("kid/a");
+                    } catch (AuthenticatorUnavailableException e) {} catch (BadCredentialsException e) {}
+                }
+            });
         }
 
-    }
-
-    static class BlockingMockKeySetProvider extends MockKeySetProvider {
-        private boolean blocked = true;
-        private boolean called = false;
+        executor.shutdown();
+        executor.awaitTermination(1, TimeUnit.SECONDS);
 
-        @Override
-        public synchronized JsonWebKeys get() throws AuthenticatorUnavailableException {
+        Assert.assertEquals(10, selfRefreshingKeySet.getRefreshCount());
+        Assert.assertEquals(0, selfRefreshingKeySet.getQueuedGetCount());
```

Comment:
Why does the count for queue become 0 here compared to 1 in the previous test?

Reply:
The previous test used Future class to call the getKey(...) method twice and the queue increments when detects a refresh Key in progress, that is why  `while (selfRefreshingKeySet.getQueuedGetCount() == 0) { Thread.sleep(10); }` was used for. So, when the system detects that the queue increments, it resumes the test and passes the assertion with number 1.
On the other hand, when I was working on improving the race condition test, there is a note in SelfRefreshingKeySetc class `// Always re-check within synchronized to handle any races` so I thought it would be more precisely if I use more threads with a lock object to access the refresh method and make sure only one thread can access the key, and then ensure all calls are executed correctly, this is why the queue must be empty at the end of the test.
@DarshitChanpura please, let me know if I answered your question correctly, and thanks for helping me to clarify my approach.

Reply:
aah, thank you for the explanation @davidosorno.